signing_service/signing_service/tests/test_authenticate.py

```diff
@@ -0,0 +1,157 @@
+
+
+class TestSigningServiceAuthenticate(SigningServiceIntegrationTestCase):
+
+    host = None
+    port = None
+    initial_reconnect_delay = None
+    ethereum_private_key = None
+    signing_service_port = None
+    authentication_challenge = None
+
+    @pytest.fixture(autouse=True)
+    def setUp(self, unused_tcp_port_factory):
+        self.host = '127.0.0.1'
+        self.port = unused_tcp_port_factory()
+        self.initial_reconnect_delay = 2
+        self.signing_service_port = unused_tcp_port_factory()
+        self.authentication_challenge = bytes(os.urandom(1000))
+
+    def test_that_authenticate_should_send_authentication_response_if_authentication_challenge_is_correct(self):
+        middleman_message = AuthenticationChallengeFrame(
+            payload=self.authentication_challenge,
+            request_id=99,
+        )
+        raw_message = middleman_message.serialize(private_key=CONCENT_PRIVATE_KEY)
+
+        raw_message_received = self._prepare_and_execute_handle_connection(
+            raw_message,
+        )
+
+        deserialized_message = AbstractFrame.deserialize(
+            raw_message=raw_message_received,
+            public_key=SIGNING_SERVICE_PUBLIC_KEY,
+        )
+
+        assertpy.assert_that(deserialized_message).is_instance_of(AuthenticationResponseFrame)
+        assertpy.assert_that(deserialized_message.payload).is_instance_of(bytes)
+        assertpy.assert_that(
+            ecdsa_verify(
+                SIGNING_SERVICE_PUBLIC_KEY,
+                deserialized_message.payload,
+                self.authentication_challenge,
+            )
+        ).is_true()
+
+    def test_that_authenticate_should_raise_socket_error_if_received_frame_is_not_authentication_challenge(self):
+        middleman_message = AuthenticationResponseFrame(
+            payload=self.authentication_challenge,
+            request_id=99,
+        )
+        raw_message = middleman_message.serialize(private_key=CONCENT_PRIVATE_KEY)
+
+        with pytest.raises(socket.error) as exception:
+            self._prepare_and_execute_handle_connection(
+                raw_message,
+            )
+
+        assertpy.assert_that(exception.value.args[0]).is_equal_to(SIGNING_SERVICE_RECOVERABLE_ERRORS[0])
+
+    def test_that_authenticate_should_raise_socket_error_if_received_frame_is_invalid(self):
+        middleman_message = AuthenticationResponseFrame(
+            payload=self.authentication_challenge,
+            request_id=99,
+        )
+        raw_message = middleman_message.serialize(private_key=CONCENT_PRIVATE_KEY)
+        malformed_raw_message = raw_message[:-1]
+
+        with pytest.raises(socket.error) as exception:
+            self._prepare_and_execute_handle_connection(
+                malformed_raw_message,
+            )
+
+        assertpy.assert_that(exception.value.args[0]).is_equal_to(SIGNING_SERVICE_RECOVERABLE_ERRORS[0])
```

Comment:
The case is that `SIGNING_SERVICE_RECOVERABLE_ERRORS` is a list of something like error codes, the exception class is always the same - `socket.error`. Also, I use `SIGNING_SERVICE_RECOVERABLE_ERRORS[0]` in code, so I have used it here too in case of if that list first element will change someday.